lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -367,6 +428,322 @@ HdVP2Mesh::HdVP2Mesh(HdVP2RenderDelegate* delegate, const SdfPath& id, const Sdf
+                    _CreateViewportCompute(*drawItem);
+#ifdef HDVP2_ENABLE_GPU_OSD
+                    _CreateOSDTables();
+#endif
+                }
+                _meshSharedData->_viewportCompute->setNormalVertexBufferGPUDirty();
+                updateNormals
+                    = false; // mark rprimDirtyBits clean? Add something to the update code?
+            } else
+#endif
+            {
+                // note: normals gets dirty when points are marked as dirty,
+                // at change tracker.
+                Hd_VertexAdjacencySharedPtr adjacency(new Hd_VertexAdjacency());
+                HdBufferSourceSharedPtr     adjacencyComputation
+                    = adjacency->GetSharedAdjacencyBuilderComputation(&_meshSharedData->_topology);
+                adjacencyComputation->Resolve();
+
+                // Only the points referenced by the topology are used to compute
+                // smooth normals.
+                VtValue normals(Hd_SmoothNormals::ComputeSmoothNormals(
+                    adjacency.get(),
+                    _points(_meshSharedData->_primvarInfo).size(),
+                    _points(_meshSharedData->_primvarInfo).cdata()));
+
+                if (!normalsInfo) {
+                    _meshSharedData->_primvarInfo[HdTokens->normals]
+                        = std::make_unique<PrimvarInfo>(
+                            PrimvarSource(normals, HdInterpolationVertex, true), nullptr);
+                } else {
+                    normalsInfo->_source.data = normals;
+                    normalsInfo->_source.interpolation = HdInterpolationVertex;
+                }
+            }
+        }
+
+        if (requireFlatNormals && (rprimDirtyBits & DirtyFlatNormals)) {
+            // TODO:
+        }
+    }
+
+    // Prepare color buffer.
+    if (((rprimDirtyBits & HdChangeTracker::DirtyPrimvar) != 0)
+        && (_PrimvarIsRequired(HdTokens->displayColor)
+            || _PrimvarIsRequired(HdTokens->displayOpacity))) {
+        HdInterpolation colorInterp = HdInterpolationConstant;
+        HdInterpolation alphaInterp = HdInterpolationConstant;
+        VtVec3fArray    colorArray;
+        VtFloatArray    alphaArray;
+
+        _getColorData(_meshSharedData->_primvarInfo, colorArray, colorInterp);
+        _getOpacityData(_meshSharedData->_primvarInfo, alphaArray, alphaInterp);
+
+        PrimvarInfo* colorAndOpacityInfo
+            = _getInfo(_meshSharedData->_primvarInfo, HdVP2Tokens->displayColorAndOpacity);
+        if (!colorAndOpacityInfo) {
+            _meshSharedData->_primvarInfo[HdVP2Tokens->displayColorAndOpacity]
+                = std::make_unique<PrimvarInfo>(
+                    PrimvarSource(VtValue(), HdInterpolationConstant, true), nullptr);
+            colorAndOpacityInfo
+                = _getInfo(_meshSharedData->_primvarInfo, HdVP2Tokens->displayColorAndOpacity);
+        }
+
+        if (!colorAndOpacityInfo->_buffer) {
```

Comment:
For color and opacity I get the buffer data & I set up the MVertexBuffer. This case is a little more complicated because color and opacity are two different primvar sources, but we merge them together into a single MVertexBuffer.

We could make the system a bit more powerful by storing more data with the primvarInfo. If we stored buffer descriptors like stride & offset then I could have two primvar sources that share an MVertexBuffer and write to different parts of it. Something to think about to clean things up further.